tests/ruby/misc_spec.rb

```diff
@@ -189,5 +189,20 @@
         expect(processes.primary.count_query("DISCARD ALL")).to eq(10)
       end
     end
+    
+    context "transaction mode with transactions" do
+      let(:processes) { Helpers::Pgcat.single_shard_setup("sharded_db", 5, "transaction") }
+      it "Does not clear set statement state when declared in a transaction" do
+        10.times do
+          conn = PG::connect(processes.pgcat.connection_string("sharded_db", "sharding_user"))
+          conn.async_exec("SET SERVER ROLE to 'primary'")
+          conn.async_exec("BEGIN")
+          conn.async_exec("SET statement_timeout to 1000")
+          conn.async_exec("COMMIT")
+          conn.close
+        end 
+        expect(processes.primary.count_query("DISCARD ALL")).to eq(0)
+      end
+    end
```

Comment:
Let's add a test for `SET LOCAL` to verify we don't send `DISCARD ALL` in these cases as well.